libcli.c

```diff
@@ -1966,11 +1990,1146 @@ void cli_set_idle_timeout_callback(struct cli_def *cli, unsigned int seconds, in
+    free_z(c);
+    c = *optarg_pair;
+  }
+}
+
+char *cli_find_optarg_value(struct cli_def *cli, char *name, char *find_after) {
+  char *value = NULL;
+  struct cli_optarg_pair *optarg_pair;
+  if (!name || !cli->found_optargs) return NULL;
+
+  for (optarg_pair = cli->found_optargs; optarg_pair && !value; optarg_pair = optarg_pair->next) {
+    if (strcmp(optarg_pair->name, name) == 0) {
+      if (find_after && (find_after == optarg_pair->value)) {
+        find_after = NULL;
+        continue;
+      }
+      value = optarg_pair->value;
+    }
+  }
+  return value;
+}
+
+static void cli_optarg_build_shortest(struct cli_optarg *optarg) {
+  struct cli_optarg *c, *p;
+  char *cp, *pp;
+  unsigned int len;
+
+  for (c = optarg; c; c = c->next) {
+    c->unique_len = 1;
+    for (p = optarg; p; p = p->next) {
+      if (c == p) continue;
+      cp = c->name;
+      pp = p->name;
+      len = 1;
+      while (*cp && *pp && *cp++ == *pp++) len++;
+      if (len > c->unique_len) c->unique_len = len;
+    }
+  }
+}
+
+void cli_free_optarg(struct cli_optarg *optarg) {
+  free_z(optarg->help);
+  free_z(optarg->name);
+  free_z(optarg);
+}
+
+int cli_register_optarg(struct cli_command *cmd, const char *name, int flags, int privilege, int mode, const char *help,
+                        int (*get_completions)(struct cli_def *cli, const char *, const char *, struct cli_comphelp *),
+                        int (*validator)(struct cli_def *cli, const char *, const char *),
+                        int (*transient_mode)(struct cli_def *cli, const char *, const char *)) {
+  struct cli_optarg *optarg;
+  struct cli_optarg *lastopt = NULL;
+  struct cli_optarg *ptr = NULL;
+  int retval = CLI_ERROR;
+
+  // name must not already exist with this priv/mode
+  for (ptr = cmd->optargs, lastopt = NULL; ptr; lastopt = ptr, ptr = ptr->next) {
+    if (!(strcmp(name, ptr->name)) && (ptr->mode == mode) && (ptr->privilege == privilege)) {
+      return CLI_ERROR;
+    }
+  }
+  if (!(optarg = calloc(sizeof(struct cli_optarg), 1))) goto CLEANUP;
+  if (!(optarg->name = strdup(name))) goto CLEANUP;
+  if (help && !(optarg->help = strdup(help))) goto CLEANUP;
+
+  optarg->mode = mode;
+  optarg->privilege = privilege;
+  optarg->get_completions = get_completions;
+  optarg->validator = validator;
+  optarg->transient_mode = transient_mode;
+  optarg->flags = flags;
+
+  if (lastopt)
+    lastopt->next = optarg;
+  else
+    cmd->optargs = optarg;
+  cli_optarg_build_shortest(cmd->optargs);
+  retval = CLI_OK;
+
+CLEANUP:
+  if (retval != CLI_OK) {
+    cli_free_optarg(optarg);
+  }
+  return retval;
+}
+
+int cli_unregister_optarg(struct cli_command *cmd, const char *name) {
+  struct cli_optarg *ptr;
+  struct cli_optarg *lastptr;
+  int retval = CLI_ERROR;
+  // iterate looking for this option name, stoping at end or if name matches
+  for (lastptr = NULL, ptr = cmd->optargs; ptr && strcmp(ptr->name, name); lastptr = ptr, ptr = ptr->next) {
+    ;
```

Comment:
Remove empty statement.

Reply:
I've got the empty statement there explicitly bracketed so show that the loop doesn't need a body - the for statement has everythign it needs.

Reply:
Sure, but the `;` inside the brackets doesn't add anything. `{}` is an empty block. In other places in this code (e.g. line 396), `;` is used on its own without the brackets, just pushing for consistency.

Reply:
Wilco.